dotnet/src/Connectors/Connectors.Memory.AzureCognitiveSearch/AzureCognitiveSearchMemory.cs

```diff
@@ -0,0 +1,316 @@
+            SemanticConfigurationName = "default",
+            QueryLanguage = "en-us", // TODO: this shouldn't be required
+            Size = limit,
+        };
+
+        Response<SearchResults<AzureCognitiveSearchRecord>>? searchResult = await client
+            .SearchAsync<AzureCognitiveSearchRecord>(query, options, cancellationToken: cancel)
+            .ConfigureAwait(false);
+
+        await foreach (SearchResult<AzureCognitiveSearchRecord>? doc in searchResult.Value.GetResultsAsync())
+        {
+            if (doc.RerankerScore < minRelevanceScore) { break; }
+
+            yield return new MemoryQueryResult(ToMemoryRecordMetadata(doc.Document), doc.RerankerScore ?? 1, null);
+        }
+    }
+
+    /// <inheritdoc />
+    public async Task RemoveAsync(string collection, string key, CancellationToken cancel = default)
+    {
+        collection = NormalizeIndexName(collection);
+
+        var records = new List<AzureCognitiveSearchRecord> { new() { Id = EncodeId(key) } };
+
+        var client = await this.GetSearchClientAsync(collection, cancel).ConfigureAwait(false);
+
+        await client.DeleteDocumentsAsync(records, cancellationToken: cancel).ConfigureAwait(false);
+    }
+
+    /// <inheritdoc />
+    public async Task<IList<string>> GetCollectionsAsync(CancellationToken cancel = default)
+    {
+        ConfiguredCancelableAsyncEnumerable<SearchIndex> indexes = this._adminClient.GetIndexesAsync(cancel).ConfigureAwait(false);
+
+        var result = new List<string>();
+        await foreach (var index in indexes)
+        {
+            result.Add(index.Name);
+        }
+
+        return result;
+    }
+
+    /// <summary>
+    /// Get a search client for the index specified.
+    /// </summary>
+    /// <param name="indexName">Index name</param>
+    /// <param name="cancellationToken">Task cancellation token</param>
+    /// <returns>Search client ready to read/write</returns>
+    private async Task<SearchClient> GetSearchClientAsync(
+        string indexName,
+        CancellationToken cancellationToken = default)
+    {
+        Response<SearchIndex>? existingIndex = null;
+        try
+        {
+            // Search the index
+            existingIndex = await this._adminClient.GetIndexAsync(indexName, cancellationToken).ConfigureAwait(false);
+        }
+        catch (RequestFailedException e) when (e.Status == 404)
+        {
+        }
+
+        // Create the index if it doesn't exist
+        if (existingIndex == null || existingIndex.Value == null)
+        {
+            await this.CreateIndexAsync(indexName, cancellationToken).ConfigureAwait(false);
+        }
+
+        // Search an available client from the local cache
+        if (!this._clientsByIndex.TryGetValue(indexName, out SearchClient? client) || client == null)
+        {
+            client = this._adminClient.GetSearchClient(indexName);
+            this._clientsByIndex[indexName] = client;
+        }
+
+        return client;
```

Comment:
Two concurrent calls to GetSearchClientAsync could end up getting a different SearchClient instance.  Is that ok?  With small tweaks here you could ensure the same instance is returned regardless of such race conditions.

Reply:
it would be fine, but I'll double check for improvements

Reply:
I can integrate the optimization later, I wouldn't block the PR on this. Could you share the code?